cmdstanpy/model.py

```diff
@@ -1378,26 +1431,53 @@ def _run_cmdstan(
     @staticmethod
     @progbar.wrap_callback
     def _wrap_sampler_progress_hook(
-        chain_id: int, total: int
+        one_process_per_chain: bool,
```

Comment:
I think we should implement this such that there is only ever one callback, it may just be called on multiple threads. It avoids the need for the if/else on the core logic.

As a sketch:
```python
    @staticmethod
    @progbar.wrap_callback
    def _wrap_sampler_progress_hook(
        chain_ids: int,
        total: int,
    ) -> Optional[Callable[[str], None]]:
        """
        Sets up tqdm callback for CmdStan sampler console msgs.
        CmdStan progress messages start with "Iteration", for single chain
        process, "Chain [id] Iteration" for multi-chain processing.
        For the latter, manage array of pbars, update accordingly.
        """
        pat = re.compile(r'Chain \[(\d*)\] (Iteration.*)')

        pbars: Dict[int, tqdm] = 
        {chain_id: tqdm(
                    total=total,
                    bar_format="{desc} |{bar}| {elapsed} {postfix[0][value]}",
                    postfix=[dict(value="Status")],
                    desc=f'chain {chain_id}',
                    colour='yellow',
            )
            for chain_id in chain_ids
        }

        def progress_hook(line: str, id: Optional[int] = None) -> None:
            if line == "Done":
                for pbar in pbars.values():
                    pbar.postfix[0]["value"] = 'Sampling completed'
                    pbar.update(total - pbar[i].n)
                    pbar.close()
            if id is None:
                match = pat.match(line)
                if match:
                    idx = int(match.group(1))
                else:
                    return # no update to do
            else:
                idx = id
                
                
            if 'Sampling' in line:
                pbar[idx].colour = 'blue'
            pbar[idx].update(1)
            pbar[idx].postfix[0]["value"] = line

        return progress_hook
```

This then lets you call either with a known id (if we're in the one-one process-chain case) or just the line and it parses it. The only other change is that this is instantiated in `sample()` instead of `run_cmdstan`, but I actually like that better (if we ever wanted a progress hook for any other cmdstan command, we'd need to do this refactor anyway)

Reply:
This is what I was working toward last week when I did some work on this PR, but I didn't get quiet to the use of the dictionary to handle chain IDs. The stuff I had with the dummy callback as the default argument should still work for `run_cmdstan` though, if we go this route

Reply:
nice - dict solves indexing problem!

Reply:
agreed this should be instantiated in sample method - at one point is was.  this code is still maddening and I'm really frustrated by the amount of logic in CmdStanPy necessary to deal with the deadly combo of CmdStan arg parser and Stan CSV files.

Reply:
I'm very much so looking forward to using something like JSON+[pydantic](https://pydantic-docs.helpmanual.io/) to do the first part if CmdStan really does move that way. The CSV part will always be a bit annoying, simply because a row-based format will never be very good at representing n-dimensional arrays.

Reply:
sorry, cannot make above sketch work.  timing myself out.

Reply:
maybe you can't update tqdm progress bars can be called from different threads?

Reply:
My implementation from last week worked in my testing with the threadprocesspool

Reply:
Want me to take a look? Waiting on some stan_math changes this week

Reply:
sure, go ahead.  it's also possible I messed up on the types or signatures.

Reply:
actually, that was exactly the problem - overlooked sig in a type hint.  fixed.  will also add debugging hooks for tqdm.

Reply:
progress bar callback hook is now instantiated in sample method and passed into `_run_cmdstan`.
not quite as fancy as your sketch, but relatively clean.